No longer possible to alias Preact?

### What's wrong?

I'm not entirely sure, but it seems like it's no longer possible to substitute Preact for React.

### Steps to reproduce

1. Add `preact` to your project
2. Alias react to `preact` according to the [documentation](https://preactjs.com/guide/v10/getting-started/#aliasing-react-to-preact):

```json
const config = { 
  "resolve": { 
    "alias": { 
      "react": "preact/compat",
      "react-dom/test-utils": "preact/test-utils",
      "react-dom": "preact/compat",
    },
  }
}
```
3. Try running cosmos

### Screenshots

![image](https://user-images.githubusercontent.com/1232595/111751167-7eb3f080-88d7-11eb-81f0-485007d3f57d.png)

### User info

- Server terminal output (see above screenshot)
- Cosmos version: 5.6.0
- Cosmos config: n/a
- If applicable: webpack version 5.26.0

### Additional context

[This comment](https://github.com/react-cosmos/react-cosmos/issues/530#issuecomment-346583016) suggests this used to work (albeit with the Preact v8 aliases).

However, as best I can tell, bce0765b280d85150abcf5a87fcdf9ef7be1f6c5 introduced a check for local react dependencies that ignores `alias` settings in the user's webpack config when looking for react.

I've made up a PR at #1317 and I've verified that at least my local cosmos setup gets past the stage where it resolves React dependencies.

Unfortunately I'm now getting errors including CSS files in my fixtures and I'm not sure if it's a result of my changes or some other changes in Cosmos over the last few weeks that broke this without me noticing.

Specifically I now get:

```log
ERROR in ./public/src/components/AsyncButton.fixture.js 4:0-22
Module not found: Error: Can't resolve '../index.css' in '/home/birtles/project/public/src/components'
 @ ./node_modules/react-cosmos/dist/plugins/webpack/client/userDeps.js 8:0-88 42:50-58
 @ ./node_modules/react-cosmos/dist/plugins/webpack/client/index.js 23:17-38 37:2-42:4

ERROR in ./public/src/components/ExpandingErrorLine.fixture.js 4:0-22
Module not found: Error: Can't resolve '../index.css' in '/home/birtles/project/public/src/components'
 @ ./node_modules/react-cosmos/dist/plugins/webpack/client/userDeps.js 9:0-95 43:57-65
 @ ./node_modules/react-cosmos/dist/plugins/webpack/client/index.js 23:17-38 37:2-42:4
...
```

The source in question is:

```ts
import '../index.css';
```

which used to work, but no longer seems to.

--- a/packages/react-cosmos/src/plugins/webpack/webpackConfig/__tests__/customDevConfig.ts
+++ b/packages/react-cosmos/src/plugins/webpack/webpackConfig/__tests__/customDevConfig.ts
@@ -128,3 +128,123 @@ it('includes HotModuleReplacementPlugin', async () => {
   );
   expect(hotModuleReplacementPlugin).toBeDefined();
 });
+
+it('applies resolve aliases for React', async () => {
+  const resolvePreact = jest.fn(() => ({
+    ...mockWebpackConfig(),
+    resolve: {
+      alias: {
+        react: 'preact/compat',
+        'react-dom': 'preact/compat',
+      },
+    },
+  }));
+  mockFile('mywebpack.config.js', resolvePreact);
+
+  const { resolve } = await getCustomDevWebpackConfig();
+  expect(resolve?.alias).toBeDefined();
+  const aliasAsObject = resolve?.alias as Record<
+    string,
+    string | false | string[]
+  >;
+  expect(aliasAsObject.react).toEqual('preact/compat');
+  expect(aliasAsObject['react-dom']).toEqual('preact/compat');
+});
+
+it('applies resolve aliases for React using exact matches', async () => {
+  const resolvePreact = jest.fn(() => ({
+    ...mockWebpackConfig(),
+    resolve: {
+      alias: {
+        react$: 'preact/compat',
+        'react-dom$': 'preact/compat',
+      },
+    },
+  }));
+  mockFile('mywebpack.config.js', resolvePreact);
+
+  const { resolve } = await getCustomDevWebpackConfig();
+  expect(resolve?.alias).toBeDefined();
+  const aliasAsObject = resolve?.alias as Record<
+    string,
+    string | false | string[]
+  >;
+  expect(aliasAsObject.react$).toEqual('preact/compat');
+  expect(aliasAsObject.react).toBeUndefined();
+  expect(aliasAsObject['react-dom$']).toEqual('preact/compat');
+  expect(aliasAsObject['react-dom']).toBeUndefined();
+});
+
+it('applies resolve aliases for React using array form', async () => {
+  const resolvePreact = jest.fn(() => ({
+    ...mockWebpackConfig(),
+    resolve: {
+      alias: [
+        {
+          name: 'react',
+          alias: 'preact/compat',
+        },
+        {
+          name: 'react-dom$',
+          alias: 'preact/compat',
+        },
+      ],
+    },
+  }));
+  mockFile('mywebpack.config.js', resolvePreact);
+
+  const { resolve } = await getCustomDevWebpackConfig();
+  expect(resolve?.alias).toBeDefined();
+  const aliasAsArray = resolve?.alias as Array<{
+    name: string;
+    alias: string | false | string[];
+  }>;
+  expect(aliasAsArray).toHaveLength(2);
+  expect(aliasAsArray).toContainEqual({
+    name: 'react',
+    alias: 'preact/compat',
+  });
+  expect(aliasAsArray).toContainEqual({
+    name: 'react-dom$',
+    alias: 'preact/compat',
+  });
+});
+
+it('adds missing React aliases by matching the alias form', async () => {
+  const resolvePreact = jest.fn(() => ({
+    ...mockWebpackConfig(),
+    resolve: {
+      alias: [
+        {
+          name: 'xyz',
+          alias: 'abc',
+        },
+      ],
+    },
+  }));
+  mockFile('mywebpack.config.js', resolvePreact);
+
+  const { resolve } = await getCustomDevWebpackConfig();
+  expect(resolve?.alias).toBeDefined();
+  const aliasAsArray = resolve?.alias as Array<{
+    name: string;
+    alias: string | false | string[];
+  }>;
+  expect(aliasAsArray).toHaveLength(3);
+  expect(aliasAsArray).toEqual(
+    expect.arrayContaining([
+      {
+        name: 'xyz',
+        alias: 'abc',
+      },
+      {
+        name: 'react',
+        alias: expect.stringMatching(/node_modules\/react$/),
+      },
+      {
+        name: 'react-dom',
+        alias: expect.stringMatching(/node_modules\/react-dom$/),
+      },
+    ])
+  );
+});

--- a/packages/react-cosmos/src/plugins/webpack/webpackConfig/shared.ts
+++ b/packages/react-cosmos/src/plugins/webpack/webpackConfig/shared.ts
@@ -87,22 +87,62 @@ export function resolveLocalReactDeps(
 ): webpack.ResolveOptions {
   const { rootDir } = cosmosConfig;
 
-  const reactPath = resolveFrom.silent(rootDir, 'react');
-  if (!reactPath) throw new Error(`[Cosmos] Local dependency not found: react`);
-
-  const reactDomPath = resolveFrom.silent(rootDir, 'react-dom');
-  if (!reactDomPath)
-    throw new Error(`[Cosmos] Local dependency not found: react-dom`);
-
+  // Fetch existing configuration
   const { resolve = {} } = baseWebpackConfig;
   const { alias = {} } = resolve;
+
+  // Check for existing React aliases
+  let hasReactAlias = false;
+  let hasReactDomAlias = false;
+  if (Array.isArray(alias)) {
+    hasReactAlias = alias.some(
+      item => item.name === 'react' || item.name === 'react$'
+    );
+    hasReactDomAlias = alias.some(
+      item => item.name === 'react-dom' || item.name === 'react-dom$'
+    );
+  } else {
+    hasReactAlias = typeof (alias.react ?? alias.react$) !== 'undefined';
+    hasReactDomAlias =
+      typeof (alias['react-dom'] ?? alias['react-dom$']) !== 'undefined';
+  }
+
+  if (hasReactAlias && hasReactDomAlias) {
+    console.log('[Cosmos] React and React DOM aliases found in Webpack config');
+    return resolve;
+  }
+
+  // Add aliases if they don't exist
+  const updatedAliases = Array.isArray(alias) ? alias.slice() : { ...alias };
+  const addAlias = (name: string, value: string | false | string[]) => {
+    if (Array.isArray(updatedAliases)) {
+      updatedAliases.push({ name, alias: value });
+    } else {
+      updatedAliases[name] = value;
+    }
+  };
+
+  if (!hasReactAlias) {
+    const reactPath = resolveFrom.silent(rootDir, 'react');
+    if (!reactPath)
+      throw new Error(`[Cosmos] Local dependency not found: react`);
+    addAlias('react', path.dirname(reactPath));
+  } else {
+    console.log('[Cosmos] React alias found in Webpack config');
+  }
+
+  if (!hasReactDomAlias) {
+    const reactDomPath = resolveFrom.silent(rootDir, 'react-dom');
+    if (!reactDomPath)
+      throw new Error(`[Cosmos] Local dependency not found: react-dom`);
+    addAlias('react-dom', path.dirname(reactDomPath));
+  } else {
+    console.log('[Cosmos] React DOM alias found in Webpack config');
+  }
+
   return {
     ...resolve,
-    alias: {
-      ...alias,
-      react: path.dirname(reactPath),
-      'react-dom': path.dirname(reactDomPath),
-    },
+    alias: updatedAliases,
   };
 }
 

